### PR TITLE
Fix DB Migration for SQLite to upgrade to 2.0

### LINKED ISSUE #13877: Upgrading 1.10 sqlite database in 2.0 fails
While it is not an important case it might be annoying to users that if they used airflow 1.10 with sqlite, the migration to 2.0 will fail on dropping constraints in `known_event` table. 

It would be great to provide some more useful message then asking the user to remove the sqlite database. 

```
[2021-01-24 08:38:42,015] {db.py:678} INFO - Creating tables
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade 03afc6b6f902 -> cf5dc11e79ad, drop_user_and_chart
Traceback (most recent call last):
  File "/Users/vijayantsoni/.virtualenvs/airflow/bin/airflow", line 11, in <module>
    sys.exit(main())
  File "/Users/vijayantsoni/.virtualenvs/airflow/lib/python3.8/site-packages/airflow/__main__.py", line 40, in main
    args.func(args)
  File "/Users/vijayantsoni/.virtualenvs/airflow/lib/python3.8/site-packages/airflow/cli/cli_parser.py", line 48, in command
    return func(*args, **kwargs)
  File "/Users/vijayantsoni/.virtualenvs/airflow/lib/python3.8/site-packages/airflow/cli/commands/db_command.py", line 31, in initdb
    db.initdb()
  File "/Users/vijayantsoni/.virtualenvs/airflow/lib/python3.8/site-packages/airflow/utils/db.py", line 549, in initdb
    upgradedb()
  File "/Users/vijayantsoni/.virtualenvs/airflow/lib/python3.8/site-packages/airflow/utils/db.py", line 688, in upgradedb
    command.upgrade(config, 'heads')
  File "/Users/vijayantsoni/.virtualenvs/airflow/lib/python3.8/site-packages/alembic/command.py", line 294, in upgrade
    script.run_env()
  File "/Users/vijayantsoni/.virtualenvs/airflow/lib/python3.8/site-packages/alembic/script/base.py", line 481, in run_env
    util.load_python_file(self.dir, "env.py")
  File "/Users/vijayantsoni/.virtualenvs/airflow/lib/python3.8/site-packages/alembic/util/pyfiles.py", line 97, in load_python_file
    module = load_module_py(module_id, path)
  File "/Users/vijayantsoni/.virtualenvs/airflow/lib/python3.8/site-packages/alembic/util/compat.py", line 182, in load_module_py
    spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/Users/vijayantsoni/.virtualenvs/airflow/lib/python3.8/site-packages/airflow/migrations/env.py", line 108, in <module>
    run_migrations_online()
  File "/Users/vijayantsoni/.virtualenvs/airflow/lib/python3.8/site-packages/airflow/migrations/env.py", line 102, in run_migrations_online
    context.run_migrations()
  File "<string>", line 8, in run_migrations
  File "/Users/vijayantsoni/.virtualenvs/airflow/lib/python3.8/site-packages/alembic/runtime/environment.py", line 813, in run_migrations
    self.get_context().run_migrations(**kw)
  File "/Users/vijayantsoni/.virtualenvs/airflow/lib/python3.8/site-packages/alembic/runtime/migration.py", line 560, in run_migrations
    step.migration_fn(**kw)
  File "/Users/vijayantsoni/.virtualenvs/airflow/lib/python3.8/site-packages/airflow/migrations/versions/cf5dc11e79ad_drop_user_and_chart.py", line 49, in upgrade
    op.drop_constraint('known_event_user_id_fkey', 'known_event')
  File "<string>", line 8, in drop_constraint
  File "<string>", line 3, in drop_constraint
  File "/Users/vijayantsoni/.virtualenvs/airflow/lib/python3.8/site-packages/alembic/operations/ops.py", line 148, in drop_constraint
    return operations.invoke(op)
  File "/Users/vijayantsoni/.virtualenvs/airflow/lib/python3.8/site-packages/alembic/operations/base.py", line 354, in invoke
    return fn(self, operation)
  File "/Users/vijayantsoni/.virtualenvs/airflow/lib/python3.8/site-packages/alembic/operations/toimpl.py", line 160, in drop_constraint
    operations.impl.drop_constraint(
  File "/Users/vijayantsoni/.virtualenvs/airflow/lib/python3.8/site-packages/alembic/ddl/sqlite.py", line 52, in drop_constraint
    raise NotImplementedError(
NotImplementedError: No support for ALTER of constraints in SQLite dialectPlease refer to the batch mode feature which allows for SQLite migrations using a copy-and-move strategy.
```

### PR DESCRIPTION
**Pull request by kaxil**
closes https://github.com/apache/airflow/issues/13877

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

**Comment by kaxil** [comment]
@ephraimbuddy @jhtimmins Can you take a look and review too please

File: airflow/migrations/versions/cf5dc11e79ad_drop_user_and_chart.py [hunk 1]
```diff
@@ -47,7 +47,12 @@ def upgrade():  # noqa: D103
 
     if 'known_event' in tables:
         for fkey in inspector.get_foreign_keys(table_name="known_event", referred_table="users"):
-            op.drop_constraint(fkey['name'], 'known_event', type_="foreignkey")
+            if conn.dialect.name == "sqlite":
+                if fkey['name']:
+                    with op.batch_alter_table(table_name='known_event') as bop:
+                        bop.drop_constraint(fkey['name'], type_="foreignkey")
+            else:
+                op.drop_constraint(fkey['name'], 'known_event', type_="foreignkey")
```
Comment (turbaszek) [inline airflow/migrations/versions/cf5dc11e79ad_drop_user_and_chart.py]: Won't `batch_alter_table` work for all? It seems to work for mysql at least:
https://github.com/apache/airflow/blob/74da0faa7b85fe8a87d0a245179f8632fc93fd51/airflow/migrations/versions/74effc47d867_change_datetime_to_datetime2_6_on_mssql_.py#L52

Comment (kaxil) [reply airflow/migrations/versions/cf5dc11e79ad_drop_user_and_chart.py]: The main reason was 
`for fkey in inspector.get_foreign_keys(table_name="known_event", referred_table="users"):` was returning 2 elements, one of which had `fkey['name']` as `None` . And if you call `bop.drop_constraint(fkey['name'], type_="foreignkey")` for sqlite with `fkey['name']` it errors and complains about name not specified while it works for MySQL and Postgres -- so thought would be easier and safe to just keep the current behaviour with Postgres and MySQL.  (didn't investigate further on the foreign key entry with None name)

Comment (ephraimbuddy) [inline airflow/migrations/versions/cf5dc11e79ad_drop_user_and_chart.py]: What if you use naming_convention? batch_alter_table supports naming_convention where unnamed foreign_keys are named during reflection. https://alembic.sqlalchemy.org/en/latest/batch.html
`naming_convention = {
    "fk":
    "fk_%(table_name)s_%(column_0_name)s_%(referred_table_name)s",
}`
```suggestion
            with op.batch_alter_table(table_name='known_event', naming_convention=naming_convention) as bop:
                bop.drop_constraint(fkey['name'], type_="foreignkey")

```

Comment (kaxil) [reply airflow/migrations/versions/cf5dc11e79ad_drop_user_and_chart.py]: Just tested it on Postgres too -- worked fine -- pushed the changes in https://github.com/apache/airflow/pull/13921/commits/890aa9336fa8d1ecf5f7305095bad44d9b961d97

Comment (kaxil) [reply airflow/migrations/versions/cf5dc11e79ad_drop_user_and_chart.py]: Removed the conditional to check DB, went with the approach Tomek suggested

**Comment by kaxil** [comment]
CI is green, ready to be re-reviewed

**Comment by jhtimmins** [comment]
@kaxil looks good! What's the reason for the context manager?

**Comment by kaxil** [comment]
> @kaxil looks good! What's the reason for the context manager?

https://alembic.sqlalchemy.org/en/latest/batch.html -- That one SQLite runs a SQL query to transfer the data from the existing table to the new one, then drop the old table

**Review by vikramkoka — APPROVED** [review]

**Pull request merged**